tpu_inference/models/jax/gpt_oss.py

```diff
@@ -0,0 +1,307 @@
+                rngs=self.rng,
+                dtype=dtype,
+                router_act='softmax',
+                random_init=self.random_init,
+                activation_ffw_td=('data', None),
+                ed_sharding=('model', None),
+                e_sharding=('model',),
+            )
+
+            moe_mlp = GptOssMoE(
+                dtype=dtype,
+                num_local_experts=num_experts,
+                hidden_size=hidden_size,
+                intermediate_size_moe=ffw_intermediate_size,
+                rngs=self.rng,
+                random_init=self.random_init,
+                router=router,
+                swiglu_limit=swiglu_limit,
+                # Sharding configuration
+                activation_ffw_td=('data', None),
+                edf_sharding=('model', None, None),
+                efd_sharding=('model', None, None),
+                ed_sharding=('model', None)
+            )
+
+            block = TransformerBlock(
+                pre_attention_norm=RMSNorm(
+                    dims=hidden_size,
+                    random_init=self.random_init,
+                    epsilon=rms_norm_eps,
+                    dtype=dtype,
+                    rngs=self.rng,
+                ),
+                pre_mlp_norm=RMSNorm(
+                    dims=hidden_size,
+                    random_init=self.random_init,
+                    epsilon=rms_norm_eps,
+                    dtype=dtype,
+                    rngs=self.rng,
+                ),
+                attn=attn,
+                custom_module=moe_mlp
+            )
+            self.layers.append(block)
+        # Note: ALL RMSNorm does not upcast input to float32, while the pytorch does
+        self.final_norm = RMSNorm(
+            dims=hidden_size,
+            rngs=self.rng,
+            random_init=self.random_init,
+            epsilon=rms_norm_eps,
+            dtype=dtype,
+        )
+
+        self.lm_head = LMhead(vocab_size=vocab_size,
+                              hidden_size=hidden_size,
+                              dtype=dtype,
+                              rngs=self.rng,
+                              vd_sharding=(('data', 'model'), None),
+                              dv_sharding=(None, ('data', 'model')),
+                              random_init=self.random_init)
+
+    # For compatibility with flax.
+    def apply(self, variables, *args, **kwargs):
+        return self.__call__(*args, **kwargs)
+
+    def load_weights(self, rng: PRNGKey, cache_dir: Optional[str] = None):
+        """Loads and transforms all weights from a checkpoint"""
+        self.rng = nnx.Rngs(rng)
+        
+        # Format: 'hf_key': ('jax_model_path', transform_function, target_shape)
+        transforms = {
+            "transpose_reshape": lambda w, shape: w.T.reshape(shape),
+            "reshape": lambda b, shape: b.reshape(shape),
+            "transpose": lambda w, _: w.T,
+        }
+
+        mappings = {
```

Comment:
this kind of string based dictionary mapping looks extremely fragile? Can you give brief summary of what it does and why it needs to be done this way?